src/Microsoft.AspNetCore.Razor.Evolution/ITagHelperBoundAttributeDescriptorBuilder.cs

```diff
@@ -0,0 +1,260 @@
+            [typeof(char).FullName] = "char",
+            [typeof(bool).FullName] = "bool",
+            [typeof(object).FullName] = "object",
+            [typeof(string).FullName] = "string",
+            [typeof(decimal).FullName] = "decimal",
+        };
+
+        private static ICollection<char> InvalidNonWhitespaceAttributeNameCharacters { get; } = new HashSet<char>(
+            new[] { '@', '!', '<', '/', '?', '[', '>', ']', '=', '"', '\'', '*' });
+
+        private bool _isEnum;
+        private string _indexerValueTypeName;
+        private string _name;
+        private string _propertyName;
+        private string _typeName;
+        private string _documentation;
+        private string _indexerNamePrefix;
+        private HashSet<RazorDiagnostic> _diagnostics;
+        private readonly string _containingTypeName;
+        private readonly Dictionary<string, string> _propertyBag;
+
+        private ITagHelperBoundAttributeDescriptorBuilder(string containingTypeName)
+        {
+            _containingTypeName = containingTypeName;
+            _propertyBag = new Dictionary<string, string>();
+        }
+
+        public static ITagHelperBoundAttributeDescriptorBuilder Create(string containingTypeName)
+        {
+            return new ITagHelperBoundAttributeDescriptorBuilder(containingTypeName);
+        }
+
+        public ITagHelperBoundAttributeDescriptorBuilder Name(string name)
+        {
+            _name = name;
+
+            return this;
+        }
+
+        public ITagHelperBoundAttributeDescriptorBuilder PropertyName(string propertyName)
+        {
+            _propertyName = propertyName;
+
+            return this;
+        }
+
+        public ITagHelperBoundAttributeDescriptorBuilder TypeName(string typeName)
+        {
+            _typeName = typeName;
+
+            return this;
+        }
+
+        public ITagHelperBoundAttributeDescriptorBuilder AsEnum()
+        {
+            _isEnum = true;
+
+            return this;
+        }
+
+        public ITagHelperBoundAttributeDescriptorBuilder AsDictionary(string attributeNamePrefix, string valueTypeName)
+        {
+            _indexerNamePrefix = attributeNamePrefix;
+            _indexerValueTypeName = valueTypeName;
+
+            return this;
+        }
+
+        public ITagHelperBoundAttributeDescriptorBuilder Documentation(string documentation)
+        {
+            _documentation = documentation;
+
+            return this;
+        }
+
+        public ITagHelperBoundAttributeDescriptorBuilder AddMetadata(string key, string value)
+        {
+            _propertyBag[key] = value;
+
+            return this;
+        }
+
+        public ITagHelperBoundAttributeDescriptorBuilder AddDiagnostic(RazorDiagnostic diagnostic)
+        {
+            EnsureDiagnostics();
+            _diagnostics.Add(diagnostic);
+
+            return this;
+        }
+
+        public IEnumerable<RazorDiagnostic> Validate()
```

Comment:
Thought about that. The reason why I veered away from that was:
- Re-using a builder becomes impossible because diagnostics accumulate.
- You're forced into validating a `TagHelperDescriptor` even if your runtime system can handle a technically "invalid" one.